🏃 Update KCP CAEP to reflect implementation changes (#2390)
kubernetes-sigs/cluster-api: docs/proposals/20191017-kubeadm-based-control-plane.md

  [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members).
- However, allow a control plane using an external etcd cluster to scale up to other numbers such as 2 or 4.
- Scale up operations must not be done in conjunction with an upgrade operation, this should be enforced using a validation webhook.
- Scale up operations are blocked based on Etcd and control plane health checks.
  - See [Healthchecks](#Healthchecks) below.

![controlplane-init-6](images/controlplane/controlplane-init-6.png)

##### Scale Down

- Scale down a stacked control plane
  - Wait for deletes to finish before scaling down
  - Find the oldest machine, or machine marked for removal and target that as the replica to remove

>   - Apply a label to the machine marked for removal (allows for reentrancy)

Reviewer: ```suggestion
  - Apply a annotation to the machine marked for removal (allows for reentrancy)
```

>   - Apply a label to the machine marked for removal (allows for reentrancy)

Reviewer 2: I think that scale down can be re-entrant if the algorithm for picking the machine to remove is deterministic. Because the current algorithm is deterministic (every Machine's creation timestamp is immutable, and the algorithm picks the oldest one), I think the the label is optional.

Reviewer: Possibly, but I'd rather be sure, especially since we need to take failure domains into account.

  - Remove the replica’s etcd member from the etcd cluster

>   - Update the kubeadm generated config map

Reviewer: We will also be adding annotations when the etcd member removal and kubeadm config modification steps are done.

  - Remove the Machine
- Scale down a control plane with an external etcd
  - Same as above minus etcd management

> - Scale down operations must not be done in conjunction with an upgrade operation, this should not impact manual operations for recovery. This should also be enforced using a validation webhook.

Reviewer: We are currently planning on enforcing this via the controller workflow. Not quite sure how we'd do it though a validation webhook at the moment.

> - Scale down operations are be blocked based on Etcd and control plane health checks

Reviewer 2: ```suggestion
- Scale down operations are blocked based on Etcd and control plane health checks.
  - If the controller has already picked a machine to scale down, the scale down operation will complete even if that machine's etcd member is not responding. This is required to make scale down re-entrant.
```

Author: I shortened that to `  - Etcd health is not required if a machine has already been selected for scale down` and moved the reentrant note to the top along with a note about annotations. Do you feel that captures the essence? I'll need to see the code before i reevaluate if that is not sufficient

Reviewer: The current in-tree code does the following:
- Marking a machine for deletion is not blocked on health checking
- Removing the etcd membership for the machine is blocked by an etcd health check failing
- Removing the config map entry is blocked by a control plane components health check failing
- Deletion of the machine is blocked by the etcd membership and config map entry being updated (enforced via annotation added after they have been completed).

##### Cluster upgrade (using create-swap-and-delete)

> - Triggered on any changes to Version or the InfrastructureTemplate

Reviewer: We now allow changes to the KubeadmConfig (with the exception of the InitConfig and ClusterConfig), though I realize now that I need to modify the hashing to accommodate for that change.

Author: we can make this update when those changes land

  - The KubeadmConfiguration, for the first iteration, is treated and validated as immutable and changes to it do not trigger an upgrade
  - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
      - By adding a timestamp field, `upgradeAfter`, which can be set to a specific time in the future

>         - Set to far future if no upgrades are desired

Reviewer: ```suggestion
        - Set to `nil` if no upgrades are desired
```

Technically it can be `nil` or the `time.Time` equivalent of 0

>         - An upgrade will run when that timestamp is passed

Reviewer: The controller will attempt to upgrade any Machines that have a CreationTimestamp that is older than `upgradeAfter`

Author: This doesn't make sense to me, can you provide an example?

If a machine was created on February 21st, 2020 and has an upgradeAfter value of February 23rd, 2020, the controller would try to upgrade that?

Reviewer: upgradeAfter is a filed on the KubeadmControlPlane

Author: that doesn't change the example. If the KubeadmControlPlane controller upgrades any control plane machine that was created before the `upgradeAfter` field, it seems to me it would create a new machine over and over until actual time has passed the `upgradeAfter` field and therefore the creationTimestamp has surpassed KubeadmControlPlane.

I may be thinking about this backwards  🤔

Reviewer: After further discussion, we do need to revisit the behavior related to the current reconciliation of the `upgradeAfter` field.

        - Good for scheduling upgrades/SLOs
        - Quickly modifying to now or yesterday if an upgrade is required immediately

> - Upgrade operations are be blocked based on Etcd and control plane health checks

Reviewer: With the current implementation in my PR, upgrade operations themselves are not blocked by the health checks, but rather the underlying scale up and scale down operations are blocked on health checks in the same way that they are when called independently. This allows for re-entrancy during the upgrade process

  - See [Healthchecks](#Healthchecks) below.

- The upgrade algorithm is as follows:
  - Find Machines that need upgrading

>     - Configuraiton hash doesn't match

Reviewer: ```suggestion
    - Configuration hash doesn't match
```

Author: Fixed a bunch of spelling errors, thanks

>     - `upgradeAfter` field is before now

Reviewer: ```suggestion
    - `creationTimestamp` field is before `upgradeAfter`
```

  - Mark machine as selected for upgrade
  - Scale up control plane

>   - Mark new machine as a new machine to replace an old machine (that has been marked for upgrade)

Reviewer: With the current PR we are adding an annotation to the Machine that is marked for replacement to say that a replacement machine has been created rather than marking the new machine.

>   - Scale down control plane

Reviewer: This should probably have a qualifier that the Machines being considered for scale down are limited to those that have been marked for replacement and marked as having a replacement machine already created. The scale down operation is further gated by the need to pass a health check prior to proceeding with the deletion.

> - Determining if a Machine is "up to date" will be done through the use of an annotation (controlplane.cluster.x-k8s.io/hash) that is placed on the Machine at creation time. The value of this annotation is generated by computing the Hash of the KubeadmControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used (when a new configuration is referenced in a new InfrastructureTemplate).

Reviewer: ```suggestion
- Determining if a Machine is "up to date" will be done through the use of a label (controlplane.cluster.x-k8s.io/hash) that is placed on the Machine at creation time. The value of this label is generated by computing the Hash of the KubeadmControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used (when a new configuration is referenced in a new InfrastructureTemplate).
```

Reviewer: Now that I think about it, we could compute the hash on the fly rather than using the label, since we're retrieving all of the owned machines, and only filtering by the label later.

Reviewer: Nevermind the above...
